From 01f977ea33b97c5a7d4dd5ea45d278abc9955c79 Mon Sep 17 00:00:00 2001
From: Govindraj Raja <Govindraj.Raja@imgtec.com>
Date: Thu, 26 Nov 2015 13:14:01 +0000
Subject: pistachio: suspend: calibrate timer1, irdiv1 and ethernet clock

This include following changes :-
1. Re-calibrate timer1 slow clock.
Timer1 block is used as rtc driver to maintain the
wall clock, its configured to run at 31.25Khz and
counterslowdiv1 is configured to 3.125Mhz since sys-pll
is scaled to 100Mhz at suspend path we need to adjust
counterslowdiv1 divider to run at 3.125Mhz which can be
done by changing divider from 112 to 32 which ensures
counterslowdiv1 is at 3.125Mhz and thus helps rtc driver
to maintain the timestamp.

2. Calibarte irdiv1 for sys_pll scale
IR can be used as a wakeup source but we need to ensure
it's at 32Khz so in case of sys-pll scale we can adjust irdiv1
and set it up back on resume.

3. Calibrate ethernet clock in suspend.
The stmm ethernet driver supports wake-on-lan(wol).
In order to support wol we need to ensure that the
ethernet clocks are correct in the suspend path while
the sys-pll is scaled.

Testing-steps:
--------------

Target:	ethtool -s eth0 wol g
Host:	wakeonlan -i <ip_addr> <mac_addr>
Target:	echo mem > /sys/power/state
Host:	wakeonlan -i <ip_addr> <mac_addr>

Change-Id: Id6ec54ac95f86b431ed1981a8f37289f68a00467
Signed-off-by: Govindraj Raja <Govindraj.Raja@imgtec.com>
---
 arch/mips/pistachio/suspend.S | 49 +++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 49 insertions(+)

diff --git a/arch/mips/pistachio/suspend.S b/arch/mips/pistachio/suspend.S
index a7a9df9..22b0056 100644
--- a/arch/mips/pistachio/suspend.S
+++ b/arch/mips/pistachio/suspend.S
@@ -43,6 +43,19 @@
 #define	CR_DDR_CTRL			0xB8149020
 #define	CR_DDR_CLK_MASK			0x00000002
 
+/* External timer dividers */
+#define	CR_TOP_COUNTERSLOWDIV1_CTRL		0xB8144914
+#define	COUNTERSLOW_DIV1_32K_DIV_VAL		0x0000006F
+#define	COUNTERSLOW_DIV1_32K_SCALED_DIV_VAL	0x0000001F
+
+#define	CR_TOP_IRDIV1_CTRL			0xB814491C
+#define	IR_DIV1_32K_DIV_VAL			0x00000057
+#define	IR_DIV1_DIV1_32K_SCALED_DIV_VAL		0x00000018
+
+#define	CR_TOP_ENETPHYCLKOUT_CTRL		0xB8144230
+#define	ENET_PHYCLKOUT_50M_DIV_VAL		0x00000006
+#define	ENET_PHYCLKOUT_50M_SCALED_DIV_VAL	0x00000001
+
 .text
 FEXPORT(pistachio_sram_suspend)
 LEAF(pistachio_sram_suspend)
@@ -101,6 +114,27 @@ sys_pll_scale:
 	or	v0, t3, t2
 	PTR_S	v0, (t0)
 
+slow_timer_clk_adj:
+	/*
+	 * Adjust external slow clock since its on sys_pll
+	 * slow_counter prediv runs at 3.125Mhz so using new divider
+	 * 0x1F (val-1) (count => 32) we can derive 3.125Mhz from 100Mhz.
+	 */
+	PTR_LI	t0, CR_TOP_COUNTERSLOWDIV1_CTRL
+	PTR_LI	t1, COUNTERSLOW_DIV1_32K_SCALED_DIV_VAL
+	PTR_S	t1, (t0)
+
+ir_adj:
+	PTR_LI	t0, CR_TOP_IRDIV1_CTRL
+	PTR_LI	t1, IR_DIV1_DIV1_32K_SCALED_DIV_VAL
+	PTR_S	t1, (t0)
+
+	/* to support wake-on-lan */
+enet_adj:
+	PTR_LI	t0, CR_TOP_ENETPHYCLKOUT_CTRL
+	PTR_LI	t1, ENET_PHYCLKOUT_50M_SCALED_DIV_VAL
+	PTR_S	t1, (t0)
+
 pll_lock_loop1:
 	PTR_LI	t0, CR_SYS_PLL_STATUS
 	PTR_L	t1, (t0)
@@ -133,6 +167,21 @@ sys_pll_setup:
 	or	v0, t3, t2
 	PTR_S	v0, (t0)
 
+slow_timer_setup:
+	PTR_LI	t0, CR_TOP_COUNTERSLOWDIV1_CTRL
+	PTR_LI	t1, COUNTERSLOW_DIV1_32K_DIV_VAL
+	PTR_S	t1, (t0)
+
+ir_setup:
+	PTR_LI	t0, CR_TOP_IRDIV1_CTRL
+	PTR_LI	t1, IR_DIV1_32K_DIV_VAL
+	PTR_S	t1, (t0)
+
+enet_setup:
+	PTR_LI	t0, CR_TOP_ENETPHYCLKOUT_CTRL
+	PTR_LI	t1, ENET_PHYCLKOUT_50M_DIV_VAL
+	PTR_S	t1, (t0)
+
 pll_lock_loop:
 	PTR_LI	t0, CR_SYS_PLL_STATUS
 	PTR_L	t1, (t0)
-- 
1.9.1

